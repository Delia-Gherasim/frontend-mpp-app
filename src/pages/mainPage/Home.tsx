import AddIcon from '@mui/icons-material/Add';
import {
    FormControl,
    InputLabel,
    MenuItem,
    Pagination,
    Select,
    SelectChangeEvent,
} from '@mui/material';
import IconButton from '@mui/material/IconButton';
import React, {useState} from 'react';
import SideBar from '../../component/SideBar';
import TopBar from '../../component/TopBar';
import {IDevice, PageEnum, dummyDeviceList} from '../../model/item.type';
import AddDevice from '../crud/Create';
import DeviceList from '../crud/DeviceList';
import EditDevice from '../crud/Update';

function Home() {
    const [deviceList, setDeviceList] = useState(dummyDeviceList as IDevice[]);
    const [shownPage, setShowPage] = useState(PageEnum.list);
    const [dataToEdit, setDataToEdit] = useState({} as IDevice);
    const [isSidebarOpen, setIsSidebarOpen] = useState(false);
    const [nr, setNr] = useState(10);
    const [page, setPage] = useState(1);

    const showListPage = () => {
        setShowPage(PageEnum.list);
    };

    const onAddDeviceHandler = () => {
        setShowPage(PageEnum.add);
    };

    const addDevice = (newItem: IDevice) => {
        setDeviceList([...deviceList, newItem]);
    };

    const deleteDevice = (data: IDevice) => {
        const indexToDelete = deviceList.indexOf(data);
        const tempList = [...deviceList];
        tempList.splice(indexToDelete, 1);
        setDeviceList(tempList);
    };

    const editDeviceData = (data: IDevice) => {
        setShowPage(PageEnum.edit);
        setDataToEdit(data);
    };

    const updateData = (data: IDevice) => {
        const filteredData = deviceList.filter((x) => x.id === data.id)[0];
        const indexOfData = deviceList.indexOf(filteredData);
        const tempData = [...deviceList];
        tempData[indexOfData] = data;
        setDeviceList(tempData);
    };

    const handleSidebarToggle = () => {
        setIsSidebarOpen(!isSidebarOpen);
    };

    const handleChange = (event: SelectChangeEvent<number>) => {
        setNr(event.target.value as number);
    };

    const handleChangePage = (
        event: React.MouseEvent<HTMLButtonElement> | null,
        newPage: number,
    ) => {
        setPage(newPage);
    };

    const displayedDevices = deviceList.slice(
        (page - 1) * nr,
        (page - 1) * nr + nr,
    );

    return (
        <div className={`container ${isSidebarOpen ? 'sidebar-open' : ''}`}>
            <TopBar />
            <SideBar
                isOpen={isSidebarOpen}
                handleToggle={handleSidebarToggle}
            />
            <section className='section-content'>
                {shownPage === PageEnum.list && (
                    <>
                        <div
                            className='title-container'
                            style={{display: 'flex', alignItems: 'center'}}
                        >
                            <h3 className='title' style={{marginRight: '20px'}}>
                                Device List:
                            </h3>
                            <IconButton
                                onClick={onAddDeviceHandler}
                                className='add-device-button'
                                style={{
                                    backgroundColor: '#DAA520',
                                    borderRadius: '5px',
                                    color: 'black',
                                    marginRight: '200px',
                                }}
                            >
                                <AddIcon />
                            </IconButton>
                            <Pagination
                                count={Math.ceil(deviceList.length / nr)}
                                page={page}
                                onChange={
                                    handleChangePage as (
                                        event: React.ChangeEvent<unknown>,
                                        page: number,
                                    ) => void
                                }
                                style={{
                                    marginTop: '0px',
                                    width: '100%',
                                    // marginRight: '10px',
                                }}
                            />
                            <FormControl
                                fullWidth
                                style={{
                                    borderRadius: '10px',
                                    borderColor: 'black',
                                    color: 'black',
                                    width: '100%',
                                    marginLeft: '0px',
                                }}
                            >
                                <InputLabel id='demo-simple-select-label'>
                                    Number of devices
                                </InputLabel>
                                <Select
                                    labelId='demo-simple-select-label'
                                    id='demo-simple-select'
                                    value={nr}
                                    label='Number of devices to display'
                                    onChange={handleChange}
                                    style={{
                                        borderRadius: '15px',
                                        borderColor: 'black',
                                        color: 'black',
                                    }}
                                >
                                    <MenuItem value={10}>10</MenuItem>
                                    <MenuItem value={20}>20</MenuItem>
                                    <MenuItem value={30}>30</MenuItem>
                                </Select>
                            </FormControl>
                        </div>
                        <DeviceList
                            list={displayedDevices}
                            onDeleteClickHandler={deleteDevice}
                            onEditClickHandler={editDeviceData}
                            numberOfItems={nr}
                        />
                    </>
                )}
                {shownPage === PageEnum.add && (
                    <AddDevice
                        onBackButtonHandle={showListPage}
                        onSaveButtonHandler={addDevice}
                    />
                )}
                {shownPage === PageEnum.edit && dataToEdit && (
                    <EditDevice
                        data={dataToEdit}
                        onBackButtonHandle={showListPage}
                        onSaveButtonHandler={updateData}
                    />
                )}
            </section>
        </div>
    );
}

export default Home;
