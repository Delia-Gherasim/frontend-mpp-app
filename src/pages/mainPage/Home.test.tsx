import {render} from '@testing-library/react';
import {describe, expect, it} from 'vitest';
import Home from './Home';

describe('Home', () => {
    it('should render the Home component', () => {
        const renderedComp = render(<Home />);
        expect(renderedComp).toMatchSnapshot();
    });
});
