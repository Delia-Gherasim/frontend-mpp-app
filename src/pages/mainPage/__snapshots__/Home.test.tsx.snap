// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home > should open the TopBar Component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div>
      <article
        class="article-header"
      >
        <header>
          <h2
            class="title"
          >
            Awesome App
          </h2>
        </header>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="article-header"
    >
      <header>
        <h2
          class="title"
        >
          Awesome App
        </h2>
      </header>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Home > should render the Home component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h3
        style="float: right;"
      >
        Sorted by Date
      </h3>
      <section
        class="section-content"
      >
        <table>
          <thead>
            <tr>
              <th>
                Type
              </th>
              <th>
                Brand
              </th>
              <th>
                Owner
              </th>
              <th>
                Accessories
              </th>
              <th>
                Warranty
              </th>
              <th>
                Date
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                laptop
              </td>
              <td>
                Asus
              </td>
              <td>
                eu
              </td>
              <td>
                Yes
              </td>
              <td>
                Yes
              </td>
              <td>
                Mon Apr 01 2024
              </td>
            </tr>
            <tr>
              <td>
                laptop8
              </td>
              <td>
                Asus
              </td>
              <td>
                eu
              </td>
              <td>
                Yes
              </td>
              <td>
                Yes
              </td>
              <td>
                Mon Apr 01 2024
              </td>
            </tr>
            <tr>
              <td>
                laptop7
              </td>
              <td>
                Asus
              </td>
              <td>
                tu
              </td>
              <td>
                No
              </td>
              <td>
                No
              </td>
              <td>
                Tue Apr 02 2024
              </td>
            </tr>
            <tr>
              <td>
                laptop2
              </td>
              <td>
                Asus
              </td>
              <td>
                eu
              </td>
              <td>
                Yes
              </td>
              <td>
                Yes
              </td>
              <td>
                Sun Apr 07 2024
              </td>
            </tr>
            <tr>
              <td>
                laptop4
              </td>
              <td>
                Asus
              </td>
              <td>
                eu
              </td>
              <td>
                Yes
              </td>
              <td>
                Yes
              </td>
              <td>
                Tue Apr 09 2024
              </td>
            </tr>
            <tr>
              <td>
                laptop9
              </td>
              <td>
                Asus
              </td>
              <td>
                tu
              </td>
              <td>
                No
              </td>
              <td>
                No
              </td>
              <td>
                Mon Apr 15 2024
              </td>
            </tr>
            <tr>
              <td>
                laptop10
              </td>
              <td>
                Asus
              </td>
              <td>
                eu
              </td>
              <td>
                Yes
              </td>
              <td>
                Yes
              </td>
              <td>
                Mon Apr 15 2024
              </td>
            </tr>
            <tr>
              <td>
                laptop11
              </td>
              <td>
                Asus
              </td>
              <td>
                tu
              </td>
              <td>
                No
              </td>
              <td>
                No
              </td>
              <td>
                Mon Apr 15 2024
              </td>
            </tr>
            <tr>
              <td>
                laptop12
              </td>
              <td>
                Asus
              </td>
              <td>
                eu
              </td>
              <td>
                Yes
              </td>
              <td>
                Yes
              </td>
              <td>
                Mon Apr 15 2024
              </td>
            </tr>
            <tr>
              <td>
                laptop13
              </td>
              <td>
                Asus
              </td>
              <td>
                tu
              </td>
              <td>
                No
              </td>
              <td>
                No
              </td>
              <td>
                Mon Apr 15 2024
              </td>
            </tr>
            <tr>
              <td>
                laptop1
              </td>
              <td>
                Asus
              </td>
              <td>
                tu
              </td>
              <td>
                No
              </td>
              <td>
                No
              </td>
              <td>
                Tue Apr 16 2024
              </td>
            </tr>
            <tr>
              <td>
                laptop3
              </td>
              <td>
                Asus
              </td>
              <td>
                tu
              </td>
              <td>
                No
              </td>
              <td>
                No
              </td>
              <td>
                Thu Apr 18 2024
              </td>
            </tr>
            <tr>
              <td>
                laptop5
              </td>
              <td>
                Asus
              </td>
              <td>
                tu
              </td>
              <td>
                No
              </td>
              <td>
                No
              </td>
              <td>
                Sat Apr 20 2024
              </td>
            </tr>
            <tr>
              <td>
                laptop6
              </td>
              <td>
                Asus
              </td>
              <td>
                eu
              </td>
              <td>
                Yes
              </td>
              <td>
                Yes
              </td>
              <td>
                Sun Apr 21 2024
              </td>
            </tr>
          </tbody>
        </table>
      </section>
    </div>
  </body>,
  "container": <div>
    <h3
      style="float: right;"
    >
      Sorted by Date
    </h3>
    <section
      class="section-content"
    >
      <table>
        <thead>
          <tr>
            <th>
              Type
            </th>
            <th>
              Brand
            </th>
            <th>
              Owner
            </th>
            <th>
              Accessories
            </th>
            <th>
              Warranty
            </th>
            <th>
              Date
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              laptop
            </td>
            <td>
              Asus
            </td>
            <td>
              eu
            </td>
            <td>
              Yes
            </td>
            <td>
              Yes
            </td>
            <td>
              Mon Apr 01 2024
            </td>
          </tr>
          <tr>
            <td>
              laptop8
            </td>
            <td>
              Asus
            </td>
            <td>
              eu
            </td>
            <td>
              Yes
            </td>
            <td>
              Yes
            </td>
            <td>
              Mon Apr 01 2024
            </td>
          </tr>
          <tr>
            <td>
              laptop7
            </td>
            <td>
              Asus
            </td>
            <td>
              tu
            </td>
            <td>
              No
            </td>
            <td>
              No
            </td>
            <td>
              Tue Apr 02 2024
            </td>
          </tr>
          <tr>
            <td>
              laptop2
            </td>
            <td>
              Asus
            </td>
            <td>
              eu
            </td>
            <td>
              Yes
            </td>
            <td>
              Yes
            </td>
            <td>
              Sun Apr 07 2024
            </td>
          </tr>
          <tr>
            <td>
              laptop4
            </td>
            <td>
              Asus
            </td>
            <td>
              eu
            </td>
            <td>
              Yes
            </td>
            <td>
              Yes
            </td>
            <td>
              Tue Apr 09 2024
            </td>
          </tr>
          <tr>
            <td>
              laptop9
            </td>
            <td>
              Asus
            </td>
            <td>
              tu
            </td>
            <td>
              No
            </td>
            <td>
              No
            </td>
            <td>
              Mon Apr 15 2024
            </td>
          </tr>
          <tr>
            <td>
              laptop10
            </td>
            <td>
              Asus
            </td>
            <td>
              eu
            </td>
            <td>
              Yes
            </td>
            <td>
              Yes
            </td>
            <td>
              Mon Apr 15 2024
            </td>
          </tr>
          <tr>
            <td>
              laptop11
            </td>
            <td>
              Asus
            </td>
            <td>
              tu
            </td>
            <td>
              No
            </td>
            <td>
              No
            </td>
            <td>
              Mon Apr 15 2024
            </td>
          </tr>
          <tr>
            <td>
              laptop12
            </td>
            <td>
              Asus
            </td>
            <td>
              eu
            </td>
            <td>
              Yes
            </td>
            <td>
              Yes
            </td>
            <td>
              Mon Apr 15 2024
            </td>
          </tr>
          <tr>
            <td>
              laptop13
            </td>
            <td>
              Asus
            </td>
            <td>
              tu
            </td>
            <td>
              No
            </td>
            <td>
              No
            </td>
            <td>
              Mon Apr 15 2024
            </td>
          </tr>
          <tr>
            <td>
              laptop1
            </td>
            <td>
              Asus
            </td>
            <td>
              tu
            </td>
            <td>
              No
            </td>
            <td>
              No
            </td>
            <td>
              Tue Apr 16 2024
            </td>
          </tr>
          <tr>
            <td>
              laptop3
            </td>
            <td>
              Asus
            </td>
            <td>
              tu
            </td>
            <td>
              No
            </td>
            <td>
              No
            </td>
            <td>
              Thu Apr 18 2024
            </td>
          </tr>
          <tr>
            <td>
              laptop5
            </td>
            <td>
              Asus
            </td>
            <td>
              tu
            </td>
            <td>
              No
            </td>
            <td>
              No
            </td>
            <td>
              Sat Apr 20 2024
            </td>
          </tr>
          <tr>
            <td>
              laptop6
            </td>
            <td>
              Asus
            </td>
            <td>
              eu
            </td>
            <td>
              Yes
            </td>
            <td>
              Yes
            </td>
            <td>
              Sun Apr 21 2024
            </td>
          </tr>
        </tbody>
      </table>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Home > should render the topbar component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div />
    <div>
      <article
        class="article-header"
      >
        <header>
          <h2
            class="title"
          >
            Awesome App
          </h2>
        </header>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="article-header"
    >
      <header>
        <h2
          class="title"
        >
          Awesome App
        </h2>
      </header>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
